internal/provider/database/resource.go

```diff
@@ -64,7 +76,44 @@ func (state *DatabaseResourceModel) Read(ctx context.Context, client openapi.Cli
 	return helper.ParseResponse(resp, state)
 }
 
-func (state *DatabaseResourceModel) Update(ctx context.Context, client openapi.ClientInterface) error {
+const (
+	DBA_PASSWORD_CHANGE_UNSUPPORTED_MSG = "Configured DBA password was changed and the server does not support updating the DBA password. " +
+		"Revert the configured DBA password to the value in the Terraform state and retry."
+)
+
+func IsDbaPasswordUpdateUnsupportedError(resp *http.Response, err error) bool {
+	if err != nil {
+		// "404 Not Found" is returned with no "detail" message if /dbaPassword	sub-resource is not supported
+		if resp.StatusCode == http.StatusNotFound {
+			if apiError, ok := err.(*helper.ApiError); !ok || apiError.GetDetail() == "" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (state *DatabaseResourceModel) Update(ctx context.Context, client openapi.ClientInterface, currentState framework.ResourceState) error {
+	// Try to update DBA password if it was changed in config
+	currentDatabase, _ := currentState.(*DatabaseResourceModel)
+	if !state.DbaPasswordMatches(currentDatabase) {
+		request := openapi.UpdateDbaPasswordModel{
```

Comment:
I think that this should block (i.e. a timeout should be used) because otherwise the new password can be accepted but eventually the rotation may fail. An async failure will be problematic because the TF state will be updated with the new password and the rotation won't be retried.

Reply:
But the password rotation command is inherently async. We basically just expose a facility for waiting for the async operation to complete, which can be invoked within the command that triggers that operation.

The question is, what does the operator do if the `nuodb-cp k8s rotate-dba-creds` command fails without reverting the state of the Kubernetes Secret, e.g. because of a transient networking issue? Does it keep retrying? In that case, there is no need for the Terraform state to retain the original password, because the operator will eventually succeed in updating the DBA password assuming that it is correct and the transient issue resolves itself.

The case where the `nuodb-cp k8s rotate-dba-creds` failure is not retriable is the one where current and target passwords are both incorrect. In that case, it does not matter which password we store in the Terraform; they are both equally useless.

Reply:
I just pushed a change that addresses the waiting issue. We still update the state before waiting and verifying that the password change was finalized, but I can't think of any reason not to do that, because like I said, the only reason why the DBA password change should fail permanently is if the original password was incorrect and therefore useless.

Reply:
> The question is, what does the operator do if the nuodb-cp k8s rotate-dba-creds command fails without reverting the state of the Kubernetes Secret, e.g. because of a transient networking issue? Does it keep retrying?

Yes, it retries.

> the only reason why the DBA password change should fail permanently is if the original password was incorrect and therefore useless.

I forgot that we give up only if the password doesn't match the SQL one. 

Regardless I think that we should support two modes:
(1) DBA password fully managed by DBaaS / TF
(2) DBA password managed by SQL

Your last change might be problematic in (2) case because any TF _apply_ will fail. Also, I think that TF must fail if the DBA password change attempt is made in case of (2). I was under the impression that this can be easily achieved if a `timeout` is used for the API call.

Reply:
The verification step that is performed only checks that the DBA password in the K8s secret matches the supplied value. The only time that we verify the DBA password against the database is when `"resync": true` is specified. I don't think (2) will be problematic unless an out-of-band call is made to `POST /dbaPassword`.

Reply:
> I don't think (2) will be problematic unless an out-of-band call is made to POST /dbaPassword.

Good point! Okay, in that can it should be fine as long as the DBA password in Kubernetes says the same as the one in TF state.